book/ecosystem.ipynb

```diff
@@ -0,0 +1,259 @@
+    "\n",
+    "**References**:\n",
+    "\n",
+    "- Temperature visualization based on https://rabernat.github.io/intro_to_physical_oceanography/02-c_ocean_temperature_salinity_stratification.html\n",
+    "- Interactive setup based on https://github.com/weiji14/30DayMapChallenge2021/blob/main/day25_interactive.py\n",
+    "- Data from the NOAA World Ocean Atlas, stored on the IRI Data Library at http://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NODC/.WOA09/."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "dbcff01f-fac0-4c1d-bb66-89dcb2e7711b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import panel as pn\n",
+    "import xarray as xr\n",
+    "import pygmt\n",
+    "pn.extension()\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7b2f9cb8-add6-45c2-9f54-8e33ed9e6f02",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Download the dataset from the IRI Data Library\n",
+    "url = 'http://iridl.ldeo.columbia.edu/SOURCES/.NOAA/.NODC/.WOA09/.Grid-1x1/.Annual/.temperature/.t_an/dods'\n",
+    "woa_temp = xr.open_dataset(url)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "b11ea510-7540-475e-9217-be4204cb1ea3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Make a panel widget for controlling the depth plotted\n",
+    "depth_slider = pn.widgets.DiscreteSlider(name='Discrete Slider', options=woa_temp.coords['depth'].values.astype(int).tolist(), value=0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3496d4db-92d9-45ee-9d48-6b4f00a19361",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Make a function for plotting the depth slice with PyGMT\n",
+    "\n",
+    "@pn.depends(depth=depth_slider)\n",
+    "def view(depth: int):\n",
+    "    fig = pygmt.Figure()\n",
+    "    pygmt.makecpt(cmap=\"vik\", series=[-2,30])\n",
+    "    fig.grdimage(grid=woa_temp.t_an.sel(depth=depth).isel(time=0), cmap=True, projection=\"R15c\", frame=True)\n",
+    "    fig.colorbar(frame=\"a5\")\n",
+    "    return fig"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d339433c-1c7a-4769-8986-edfb2a9897ed",
+   "metadata": {},
+   "source": [
+    "### Make the interactive dashboard!\n",
+    "\n",
+    "Now to put everything together! The 'dashboard' will be very simple.\n",
+    "The 'depth' slider is placed next to the map using `panel.Column`.\n",
+    "Selecting different depths will update the data plotted! Find out more at\n",
+    "https://panel.holoviz.org/getting_started/index.html#using-panel.\n",
+    "\n",
+    "Note: This is meant to run in a Jupyter lab/notebook environment.\n",
+    "The grdinfo warning can be ignored."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cc8d2f88-6e37-4ca4-9ebc-4bdf1e00ec8e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pn.Column(depth_slider, view)"
```

Comment:
> Note, I might need to find a different dataset to use for the interactive component. It previously was working great, but now oddly all the depth levels seem to contain the same data.

I'm also having trouble getting the plot to change with the slider in my local Jupyter notebook. Are you sure it's the data, or something up with `panel`?

Another thing is with rendering the Panel plot in JupyterBook. There's supposed to be a way to get interactivity via `thebe` (see https://jupyterbook.org/en/stable/interactive/thebe.html), but not sure how it works in practice. You could play with the config and see how it goes (no rush though, you can do this in a follow-up PR after the recording).

P.S. I'm also trying to get `pygmt` working in `pyodide` at https://github.com/pyodide/pyodide/pull/2525, but don't count on that getting finished anytime soon.

Reply:
Yes, I think so because np.testing.assert_allclose on the values from two depth slices didn't report any differences and I got the same result with a quick matplotlib plot of different depths. I'll try to figure it out tomorrow, but wanted to open the PR in case you and Leo disagreed with the general idea for the content.

> P.S. I'm also trying to get pygmt working in pyodide at https://github.com/pyodide/pyodide/pull/2525, but don't count on that getting finished anytime soon.

Awesome, super excited to see the outcome!

Reply:
Weirdly enough, it works fine when reading from disk as a netcdf. It's not a PyGMT/GMT problem, something to do with either the IRI store or OPeNDAP. I changed it to download using pygmt.which first.